### PR TITLE
Implement xwayland primary selection sync

### LINKED ISSUE #513: sync primary selection with X

### PR DESCRIPTION
**Pull request by emersion**
Test plan:
* Check that copying between wayland and xwayland clients still works
* Select some text in wayland, press middle button in xwayland
* Same but swap wayland and xwayland

Fixes #513

### CODE REVIEW

**Comment by SirCmpwn** [comment]
I would test this but it'll take a bit of work to get GTK+ stuff working on my laptop right now.

**Comment by martinetd** [comment]
Getting a use-after-free ASAN error when selecting text in gnome-terminal after pasting from xterm (I still need to figure why urxvt is tiineey on X11 backend...)

```
==9657==ERROR: AddressSanitizer: heap-use-after-free on address 0x60c000122208 at pc 0x7f1619744370 bp 0x7fff6b1dad70 sp 0x7fff6b1dad68
READ of size 8 at 0x60c000122208 thread T0
    #0 0x7f161974436f in offer_resource_handle_destroy ../types/wlr_primary_selection.c:54
    #1 0x7f16194a6f2f  (/usr/lib/x86_64-linux-gnu/libwayland-server.so.0+0x7f2f)
    #2 0x7f16194a6f81 in wl_resource_destroy (/usr/lib/x86_64-linux-gnu/libwayland-server.so.0+0x7f81)
    #3 0x7f16197442bc in offer_handle_destroy ../types/wlr_primary_selection.c:36
    #4 0x7f1615684fcd in ffi_call_unix64 (/usr/lib/x86_64-linux-gnu/libffi.so.6+0x5fcd)
    #5 0x7f161568493e in ffi_call (/usr/lib/x86_64-linux-gnu/libffi.so.6+0x593e)
    #6 0x7f16194aaa3a  (/usr/lib/x86_64-linux-gnu/libwayland-server.so.0+0xba3a)
    #7 0x7f16194a727e  (/usr/lib/x86_64-linux-gnu/libwayland-server.so.0+0x827e)
    #8 0x7f16194a8c11 in wl_event_loop_dispatch (/usr/lib/x86_64-linux-gnu/libwayland-server.so.0+0x9c11)
    #9 0x7f16194a7449 in wl_display_run (/usr/lib/x86_64-linux-gnu/libwayland-server.so.0+0x8449)
    #10 0x563fb7704f29 in main ../rootston/main.c:82
    #11 0x7f16184d9560 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x20560)
    #12 0x563fb76f3d19 in _start (/home/shared/wayland/wlroots/build/rootston/rootston+0x8d19)

0x60c000122208 is located 8 bytes inside of 120-byte region [0x60c000122200,0x60c000122278)
freed by thread T0 here:
    #0 0x7f1619a7f8c8 in __interceptor_free (/usr/lib/x86_64-linux-gnu/libasan.so.4+0xd98c8)
    #1 0x7f16196e4da4 in primary_selection_source_cancel ../xwayland/selection.c:537
    #2 0x7f1619744f94 in wlr_seat_set_primary_selection ../types/wlr_primary_selection.c:211
    #3 0x7f1619745190 in device_handle_set_selection ../types/wlr_primary_selection.c:247
    #4 0x7f1615684fcd in ffi_call_unix64 (/usr/lib/x86_64-linux-gnu/libffi.so.6+0x5fcd)

previously allocated by thread T0 here:
    #0 0x7f1619a7fdf8 in calloc (/usr/lib/x86_64-linux-gnu/libasan.so.4+0xd9df8)
    #1 0x7f16196e595f in xwm_selection_get_targets ../xwayland/selection.c:643
    #2 0x7f16196e5d05 in xwm_handle_selection_notify ../xwayland/selection.c:688
    #3 0x7f16196e644d in xwm_handle_selection_event ../xwayland/selection.c:755
    #4 0x7f16196f173b in x11_event_handler ../xwayland/xwm.c:955
    #5 0x7f16194a8c67 in wl_event_loop_dispatch (/usr/lib/x86_64-linux-gnu/libwayland-server.so.0+0x9c67)

SUMMARY: AddressSanitizer: heap-use-after-free ../types/wlr_primary_selection.c:54 in offer_resource_handle_destroy
Shadow bytes around the buggy address:
  0x0c188001c3f0: fa fa fa fa fa fa fa fa fd fd fd fd fd fd fd fd
  0x0c188001c400: fd fd fd fd fd fd fd fd fa fa fa fa fa fa fa fa
  0x0c188001c410: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c188001c420: fa fa fa fa fa fa fa fa fd fd fd fd fd fd fd fd
  0x0c188001c430: fd fd fd fd fd fd fd fd fa fa fa fa fa fa fa fa
=>0x0c188001c440: fd[fd]fd fd fd fd fd fd fd fd fd fd fd fd fd fa
  0x0c188001c450: fa fa fa fa fa fa fa fa 00 00 00 00 00 00 00 00
  0x0c188001c460: 00 00 00 00 00 00 00 00 fa fa fa fa fa fa fa fa
  0x0c188001c470: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c188001c480: fa fa fa fa fa fa fa fa fd fd fd fd fd fd fd fd
  0x0c188001c490: fd fd fd fd fd fd fd fd fa fa fa fa fa fa fa fa
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
==9657==ABORTING
```

**Comment by emersion** [comment]
Thanks for the heads up, @martinetd! This should be fixed with the last commit, can you confirm?

**Comment by martinetd** [comment]
Cheers, I can confirm the use-after-free is gone so can test a bit further.
It looks like the sync from Xwayland to xdg_shell works, but not the other way around. I get the following message when trying to paste to an Xwayland window. Reproduced with drm backend as well as X11's.
```
2017-12-25 18:42:26 - [xwayland/selection.c:314] not handling selection request: no selection source assigned to xwayland seat
```

**Comment by emersion** [comment]
Oh, that's an old check that I forgot to remove. Should be fixed now!

File: xwayland/selection.c [hunk 1]
```diff
@@ -652,52 +804,69 @@ void xwm_selection_init(struct wlr_xwm *xwm) {
...
-	xcb_set_selection_owner(xwm->xcb_conn,
-		xwm->selection_window,
-		xwm->atoms[CLIPBOARD],
-		XCB_TIME_CURRENT_TIME);
+	xwm_selection_set_owner(&xwm->primary_selection, source != NULL);
 }
 
 void xwm_set_seat(struct wlr_xwm *xwm, struct wlr_seat *seat) {
 	assert(xwm);
 	assert(seat);
```
Comment (acrisci) [inline xwayland/selection.c]: This doesn't handle unsetting the seat.

File: xwayland/selection.c [hunk 2]
```diff
@@ -511,33 +596,83 @@ static void xwm_get_selection_targets(struct wlr_xwm *xwm) {
...
+		source->base.cancel = data_source_cancel;
+
+		source->selection = selection;
+		wl_array_init(&source->mime_types_atoms);
+
+		bool ok = source_get_targets(selection, &source->base.mime_types,
+			&source->mime_types_atoms);
+		if (ok) {
+			wlr_seat_set_selection(xwm->seat, &source->base,
+				wl_display_next_serial(xwm->xwayland->wl_display));
```
Comment (acrisci) [inline xwayland/selection.c]: Setting the selection like this with a generated serial will cause some problems with #511. There's no way for the seat to internally verify this serial. We'll probably need to have a way to add serials to the validation list that we know to be valid internally.

**Comment by acrisci** [comment]
Seems to work ok.

Can you add a log message for each of the xcb events you handle?

**Comment by acrisci** [comment]
👍

**Comment by SirCmpwn** [comment]
Tested this, any selection originating from gnome-terminal is not usable anywhere else. All other test cases work (mainly using gedit and urxvt).

**Pull request merged**